src/js/controller/controller.js

```diff
@@ -280,33 +274,119 @@ define([
+                    var iFrame = _model.get('iFrame');
+                    if (iFrame) {
+                        return 1;
+                    }
+
+                    // Get intersectionRatio through brute force
+                    intersectionRatio = _computeVisibility(_view.element());
                 }
 
-                var visibility = activeTab ? intersectionRatio : 0;
-                _model.set('visibility', visibility);
+                return intersectionRatio;
+            }
+
+            function _computeVisibility(target) {
+                var html = document.documentElement;
+                var body = document.body;
+                var rootRect = {
+                    top: 0,
+                    left: 0,
+                    right: html.clientWidth || body.clientWidth,
+                    width: html.clientWidth || body.clientWidth,
+                    bottom: html.clientHeight || body.clientHeight,
+                    height: html.clientHeight || body.clientHeight
+                };
+
+                var targetRect = utils.bounds(target);
+                var intersectionRect = targetRect;
+                var parent = target.parentNode;
+                var atRoot = false;
+
+                while (!atRoot) {
+                    var parentRect = null;
+                    if (!parent || parent.nodeType !== 1) {
+                        atRoot = true;
+                        parentRect = rootRect;
+                    } else if (window.getComputedStyle(parent).overflow !== 'visible') {
+                        parentRect = utils.bounds(parent);
+                    }
+                    if (parentRect) {
+                        intersectionRect = computeRectIntersection(parentRect, intersectionRect);
+                        if (!intersectionRect) {
+                            return 0;
+                        }
+                    }
+                    parent = parent.parentNode;
+                }
+                var targetArea = targetRect.width * targetRect.height;
+                var intersectionArea = intersectionRect.width * intersectionRect.height;
+                return targetArea ? (intersectionArea / targetArea) : 0;
+            }
+
+            function computeRectIntersection(rect1, rect2) {
+                var top = Math.max(rect1.top, rect2.top);
+                var bottom = Math.min(rect1.bottom, rect2.bottom);
+                var left = Math.max(rect1.left, rect2.left);
+                var right = Math.min(rect1.right, rect2.right);
+                var width = right - left;
+                var height = bottom - top;
+                return (width >= 0 && height >= 0) && {
+                    top: top,
+                    bottom: bottom,
+                    left: left,
+                    right: right,
+                    width: width,
+                    height: height
+                };
+            }
+
+            function _setViewable(visibility) {
```

Comment:
It simplifies `_onVisibilityChange` so we don't have to read off the model after setting it.